api/v1/storageprofile_types.go

```diff
@@ -24,25 +24,25 @@ import (
 // NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
 
 // StorageProfileSpec defines the desired state of StorageProfile
-// +kubebuilder:validation:XValidation:rule="!has(oldSelf.deviceClass) || has(self.deviceClass)", message="deviceClass is required once set"
-// +kubebuilder:validation:XValidation:rule="!has(oldSelf.sharedFilesystemConfiguration) || has(self.sharedFilesystemConfiguration)", message="sharedFilesystemConfiguration is required once set"
-// +kubebuilder:validation:XValidation:rule="!has(oldSelf.blockPoolConfiguration) || has(self.blockPoolConfiguration)", message="blockPoolConfiguration is required once set"
+// +kubebuilder:validation:XValidation:rule="has(oldSelf.deviceClass) && has(self.deviceClass)", message="deviceClass is required"
+// +kubebuilder:validation:XValidation:rule="has(oldSelf.sharedFilesystemConfiguration) && has(self.sharedFilesystemConfiguration)", message="sharedFilesystemConfiguration is required"
+// +kubebuilder:validation:XValidation:rule="has(oldSelf.blockPoolConfiguration) && has(self.blockPoolConfiguration)", message="blockPoolConfiguration is required"
```

Comment:
i think the earlier check was correct, any specific reason to change it?

Reply:
The old rule applied to a different use case, as mentioned in the previous PR [comment](https://github.com/red-hat-storage/ocs-operator/pull/2204#discussion_r1356840478)

Reply:
- I agree with Madhu, `deviceClass` is required and immutable, everything else is optional but required and immutable after once set.
- btw, current rule on type states, during a transition existing (self) value and new (oldSelf) value should be set which is already achieved with `// +kubebuilder:validation:Required` and so these two rules are redundant
- Only question is, can a complex field be `Immutable once set after modification` as mentioned in another comment (without any side effects to shared CRDs)?

Reply:
@leelavg 

1. No one said otherwise.
2. As mentioned above, I am testing it with new not yet committed changes.
3. It doesn't matter, we aim to eliminate this type of behavior, I was only mentioning it to prove my case to Ohad's point that it can be used on a per _non_ sub-field basis.